src/box/field_def.c

```diff
@@ -155,6 +242,19 @@ field_type1_contains_type2(enum field_type type1, enum field_type type2)
 	assert(type2 < field_type_MAX);
 	if (type1 == type2)
 		return true;
+	/*
+	 * In general, all values of `type2' are forbidden for the fixed-size
+	 * integer field `type1'. Such values require a range check.
+	 */
+	bool is_fixed_int = type1 == FIELD_TYPE_INT8 ||
+			    type1 == FIELD_TYPE_UINT8 ||
+			    type1 == FIELD_TYPE_INT16 ||
+			    type1 == FIELD_TYPE_UINT16 ||
+			    type1 == FIELD_TYPE_INT32 ||
+			    type1 == FIELD_TYPE_UINT32 ||
+			    type1 == FIELD_TYPE_INT64;
+	if (is_fixed_int)
+		return false;
```

Comment:
It should be possible to convert `uint16` to `uint32` or `uint64` without checking values. Same for signed integer types and, probably, even `float32` -> `float64`.